solr/core/src/java/org/apache/solr/response/transform/ChildDocTransformer.java

```diff
@@ -0,0 +1,244 @@
+      }
+    } catch (IOException e) {
+      rootDoc.put(getName(), "Could not fetch child Documents");
+    }
+  }
+
+  private void addChildrenToParent(SolrDocument parent, Multimap<String, SolrDocument> children) {
+    for(String childLabel: children.keySet()) {
+      addChildrenToParent(parent, children.get(childLabel), childLabel);
+    }
+  }
+
+  private void addChildrenToParent(SolrDocument parent, Collection<SolrDocument> children, String cDocsPath) {
+    // if anonChildDoc is set to true we do not need to add the child document's relation to its parent document.
+    if(anonChildDoc) {
+      parent.addChildDocuments(children);
+      return;
+    }
+    // lookup leaf key for these children using path
+    // depending on the label, add to the parent at the right key/label
+    String trimmedPath = trimLastPound(cDocsPath);
+    // if the child doc's path does not end with #, it is an array(same string is returned by ChildDocTransformer#trimLastPound)
+    if (!parent.containsKey(trimmedPath) && (trimmedPath == cDocsPath)) {
+      List<SolrDocument> list = new ArrayList<>(children);
+      parent.setField(trimmedPath, list);
+      return;
+    }
+    // is single value
+    parent.setField(trimmedPath, ((List)children).get(0));
+  }
+
+  private static String getLastPath(String path) {
+    int lastIndexOfPathSepChar = path.lastIndexOf(PATH_SEP_CHAR.charAt(0));
+    if(lastIndexOfPathSepChar == -1) {
+      return path;
+    }
+    return path.substring(lastIndexOfPathSepChar + 1);
+  }
+
+  private static String trimPathIfArrayDoc(String path) {
+    // remove index after last pound sign and if there is an array index e.g. toppings#1 -> toppings
+    // or return original string if child doc is not in an array ingredients# -> ingredients#
+    int lastIndex = path.length() - 1;
+    boolean singleDocVal = path.charAt(lastIndex) == NUM_SEP_CHAR.charAt(0);
+    return singleDocVal ? path: path.substring(0, path.lastIndexOf(NUM_SEP_CHAR.charAt(0)));
+  }
+
+  private static String trimLastPound(String path) {
+    // remove index after last pound sign and index from e.g. toppings#1 -> toppings
+    int lastIndex = path.lastIndexOf('#');
+    return lastIndex == -1 ? path: path.substring(0, lastIndex);
+  }
+
+  /**
+   * Returns the *parent* path for this document.
+   * Children of the root will yield null.
+   */
+  private static String getParentPath(String currDocPath) {
+    // chop off leaf (after last '/')
+    // if child of leaf then return null (special value)
+    int lastPathIndex = currDocPath.lastIndexOf(PATH_SEP_CHAR);
+    return lastPathIndex == -1 ? null: currDocPath.substring(0, lastPathIndex);
+  }
+
+  private static String getPathByDocId(int segDocId, SortedDocValues segPathDocValues) throws IOException {
+    int numToAdvance = segPathDocValues.docID()==-1?segDocId: segDocId - (segPathDocValues.docID());
+    assert numToAdvance >= 0;
+    assert segPathDocValues.advanceExact(segDocId);
```

Comment:
Woah; we want to call advanceExact even if assertions are disabled -- like, you know, in production/real-world; not just tests.  Thankfully, Lucene/Solr's randomized test infrastructure actually randomly disables assertions and thus would catch this (eventually) if I didn't.